ksql-engine/src/main/java/io/confluent/ksql/topic/DefaultTopicInjector.java

```diff
@@ -0,0 +1,202 @@
+      final KsqlConfig ksqlConfig,
+      final Map<String, Object> propertyOverrides) {
+    if (!(statement.getStatement() instanceof CreateAsSelect)) {
+      return statement;
+    }
+
+    final PreparedStatement<? extends CreateAsSelect> cas =
+        (PreparedStatement<? extends CreateAsSelect>) statement;
+
+    final String topic = topicName(cas.getStatement(), ksqlConfig);
+    final TopicDescription description = describeSource(topicClient, ksqlConfig, cas);
+    final int partitions = numPartitions(cas, ksqlConfig, propertyOverrides, description);
+    final short replicas = numReplicas(cas, ksqlConfig, propertyOverrides, description);
+
+    final Map<String, Expression> properties = new HashMap<>(cas.getStatement().getProperties());
+    properties.put(DdlConfig.KAFKA_TOPIC_NAME_PROPERTY, new StringLiteral(topic));
+    properties.put(KsqlConstants.SINK_NUMBER_OF_REPLICAS, new IntegerLiteral(replicas));
+    properties.put(KsqlConstants.SINK_NUMBER_OF_PARTITIONS, new IntegerLiteral(partitions));
+
+    final CreateAsSelect withTopic = cas.getStatement().copyWith(properties);
+    final String withTopicText = SqlFormatter.formatSql(withTopic) + ";";
+
+    return (PreparedStatement<T>) PreparedStatement.of(withTopicText, withTopic);
+  }
+
+  private TopicDescription describeSource(
+      final KafkaTopicClient topicClient,
+      final KsqlConfig ksqlConfig,
+      final PreparedStatement<? extends CreateAsSelect> cas
+  ) {
+    final Analysis analysis = new QueryAnalyzer(
+        metaStore,
+        ksqlConfig.getString(KsqlConfig.KSQL_OUTPUT_TOPIC_NAME_PREFIX_CONFIG))
+        .analyze(
+            cas.getStatementText(),
+            cas.getStatement().getQuery(),
+            Optional.of(cas.getStatement().getSink()));
+
+    final StructuredDataSource theSource = analysis.getTheSource();
+    final String kafkaTopicName = theSource.getKsqlTopic().getKafkaTopicName();
+    return topicClient.describeTopic(kafkaTopicName);
+  }
+
+  private static String topicName(
+      final CreateAsSelect cas,
+      final KsqlConfig cfg
+  ) {
+    final Expression topicProperty = cas.getProperties().get(DdlConfig.KAFKA_TOPIC_NAME_PROPERTY);
+    return (topicProperty == null)
+        ? cfg.getString(KsqlConfig.KSQL_OUTPUT_TOPIC_NAME_PREFIX_CONFIG) + cas.getName().getSuffix()
+        : StringUtils.strip(topicProperty.toString(), "'");
+  }
+
+  private static int numPartitions(
+      final PreparedStatement<? extends CreateAsSelect> cas,
+      final KsqlConfig cfg,
+      final Map<String, Object> propertyOverrides,
+      final TopicDescription description
+  ) {
+    final Optional<Integer> inProperties = checkedParse(
+        Integer::parseInt, cas, KsqlConstants.SINK_NUMBER_OF_PARTITIONS);
+    if (inProperties.isPresent()) {
+      return inProperties.get();
+    }
+
+    final Integer legacyNumPartitions = cfg.getInt(KsqlConfig.SINK_NUMBER_OF_PARTITIONS_PROPERTY);
+    return checkedParse(
+          Integer::parseInt,
```

Comment:
Minor edge case, but I think parseInt does not fail on overflow... 

same for parseShort.

Reply:
It does fail :)  added test